infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/LRRollingUpgradeHandler.java

```diff
@@ -0,0 +1,60 @@
+package org.corfudb.infrastructure.logreplication.infrastructure;
+
+import org.corfudb.infrastructure.logreplication.infrastructure.plugins.ILogReplicationVersionAdapter;
+import org.corfudb.runtime.collections.TxnContext;
+
+/**
+ * Rolling upgrade handling means cluster must function in a mode where not all
+ * nodes are running the same codebase.
+ * Here the newer code base must function in a backward compatible mode until all
+ * the nodes have been upgraded to the same version.
+ * Once all nodes have upgraded it may atomically migrate data, clean up old code and switch
+ * completely to newer logic.
+ *
+ * To facilitate this in LR we add this module which is intended to be used as follows
+ * on LR startup and when upgrade is complete.
+ *
+ * For anything post corfu-0.4.0.1 we consider the code base to be at "V2"
+ * Practically this means any code that is running Active/Standby logic is considered "V1"
+ * Any code shipped that uses Source/Sink can be thought of as "V2"
+ *                  ┌─────────────────────────────────┐
+ *                  │ isLRReadyForV2 = false (in-mem) │
+ *                  └──────────────────┬──────────────┘
+ *                                     │
+ *                     ┌───────────────▼───────────┐
+ *                   ┌─┤                           ├─┐
+ *         Yes ┌─────┴─┴┐  isRollingUpgradeOn() ? ┌┴─┴──────┐ No
+ *             │        └─────────────────────────┘         │
+ *             │                                            │
+ * ┌───────────▼────────────┐                         ┌──────▼────────────────┐
+ * │ isLRReadyForV2 = false │                         │ isLRReadyForV2 = true │
+ * └───────────┬────────────┘                         └───────┬───────────────┘
+ *             │            ┌──────────────────────┐          │
+ *             └───────────►│ start Log Replicator │◄─────────┘
+ *                          └──────────────────────┘
+ */
+public class LRRollingUpgradeHandler {
+
+    private volatile boolean isLRReadyForV2 = false;
```

Comment:
would it be better to store this information persistently so that the check need not be performed every time LR starts?  We can do it in another PR...

Reply:
Kind of done in the re-work. thanks